README.md
# kzdroe

Here is a simple Python Flask for receiving a recording from doe.dialbox.cloud.

# Build And Push to docker registry

``` Bash
docker build -t sapian/kzdroe:0.2.1 -t sapian/kzdroe:latest -t us.gcr.io/sapiancomco-1531131691088/sapian/kzdroe:0.2.1 .
```

# Run For Testing

## Set Enviroment variables
Fill de file `.env`

``` ini
LOG_LEVEL=INFO
```

## Run

``` Bash
docker run --env-file .env --rm -it sapian/kzdroe:latest
```


CHANGELOG.md
# 0.3.2

* BETTER start_call time detection

# 0.3.1

* ADDED s3 uploadfile

# 0.2.1

* ADDED opus Encode 

# 0.1.1

* First Verision 

kzdroe/kzdroe/s3.py
#!/usr/bin/env python
# vim: set fileencoding=utf-8 :
#
#       kzdroe/kzdroe/s3.py
#       Copyright 2021 sebastian.rojo <sebastian.rojo@sapian.com.co>
#
#       This program is free software; you can redistribute it and/or modify
#       it under the terms of the GNU General Public License as published by
#       the Free Software Foundation; either version 2 of the License, or
#       (at your option) any later version.
#
#       This program is distributed in the hope that it will be useful,
#       but WITHOUT ANY WARRANTY; without even the implied warranty of
#       MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#       GNU General Public License for more details.
#
#       You should have received a copy of the GNU General Public License
#       along with this program; if not, write to the Free Software
#       Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
#       MA 02110-1301, USA.
#

__author__ = "Sebastian Rojo"
__copyright__ = "Copyright 2020, Sebastian Rojo, Sapian SAS"
__credits__ = []
__license__ = "GPLv2"
__version__ = "0.1.1"
__maintainer__ = "Sebastian Rojo"
__email__ = ["sebastian.rojo at sapian.com.co", "arpagon at gmail.com"]
__status__ = "Beta"

import os
import logging
import boto3
import botocore

LOG_LEVEL = os.environ.get('LOG_LEVEL', "INFO").upper()
S3 = int(os.environ.get('S3', False))
logging.basicConfig(level=LOG_LEVEL)
log = logging.getLogger('KZDROE.S3')

if S3:
    S3_ACCESS_KEY = os.environ.get('S3_ACCESS_KEY', "")
    S3_SECRET_KEY = os.environ.get('S3_SECRET_KEY', "")
    S3_HOST = os.environ.get('S3_HOST', "")

s3 = boto3.client(
        's3',
        endpoint_url=S3_HOST,
        aws_access_key_id=S3_ACCESS_KEY,
        aws_secret_access_key=S3_SECRET_KEY,
        region_name=''
    )

s3_resource = boto3.resource(
        's3',
        endpoint_url=S3_HOST,
        aws_access_key_id=S3_ACCESS_KEY,
        aws_secret_access_key=S3_SECRET_KEY,
        region_name=''
    )


def check_bucket(bucket_name):
    try:
        log.debug("")
        s3_resource.meta.client.head_bucket(Bucket=bucket_name)
        log.debug("Bucket Exists!")
        return True
    except botocore.exceptions.ClientError as e:
        # If a client error is thrown, then check that it was a 404 error.
        # If it was a 404 error, then the bucket does not exist.
        error_code = int(e.response['Error']['Code'])
        if error_code == 403:
            log.debug("Private Bucket. Forbidden Access!")
            return True
        elif error_code == 404:
            log.debug("Bucket Does Not Exist!")
            return False

def upload_file(file_name, bucket_name, object_name=None):
    """Upload a file to an S3 bucket

    :param file_name: File to upload
    :param bucket: Bucket to upload to
    :param object_name: S3 object name. If not specified then file_name is used
    :return: True if file was uploaded, else False
    """
    # check if bucket exist
    if not check_bucket(bucket_name):
        s3.create_bucket(Bucket=bucket_name)

    # If S3 object_name was not specified, use file_name
    if object_name is None:
        object_name = file_name

    # Upload the file    
    try:
        response = s3.upload_file(file_name, bucket_name, object_name)
    except botocore.exceptions.ClientError as e:
        log.error(e)
        return False
    return True

if __name__ == '__main__':
    if not (S3_ACCESS_KEY and S3_SECRET_KEY):
        log.error(
            "ERROR: for S3 storage need S3_ACCESS_KEY and S3_SECRET_KEY"
        )
        exit(1)

kzdroe/kzdroe/app.py
#!/usr/bin/env python
# vim: set fileencoding=utf-8 :
#
#       kzdroe/kzdroe/app.py
#       Copyright 2020 sebastian.rojo <sebastian.rojo@sapian.com.co>
#
#       This program is free software; you can redistribute it and/or modify
#       it under the terms of the GNU General Public License as published by
#       the Free Software Foundation; either version 2 of the License, or
#       (at your option) any later version.
#
#       This program is distributed in the hope that it will be useful,
#       but WITHOUT ANY WARRANTY; without even the implied warranty of
#       MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#       GNU General Public License for more details.
#
#       You should have received a copy of the GNU General Public License
#       along with this program; if not, write to the Free Software
#       Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
#       MA 02110-1301, USA.
#

__author__ = "Sebastian Rojo"
__copyright__ = "Copyright 2020, Sebastian Rojo, Sapian SAS"
__credits__ = []
__license__ = "GPLv2"
__version__ = "0.1.1"
__maintainer__ = "Sebastian Rojo"
__email__ = ["sebastian.rojo at sapian.com.co", "arpagon at gmail.com"]
__status__ = "Beta"

import os
import logging
from datetime import datetime
from flask import Flask
from flask import request
from flask import jsonify
from prometheus_flask_exporter import PrometheusMetrics
from prometheus_flask_exporter.multiprocess import UWsgiPrometheusMetrics
from shutil import copyfile, which
from subprocess import Popen

SOURCE_AUDIO_EXTENSIONS = ['.wav']
GREGORIAN_SECONDS=62167219200
LOG_LEVEL = os.environ.get('LOG_LEVEL', "INFO").upper()
METRICS_PORT = int(os.environ.get('METRICS_PORT', "9531"))
ENCOPUS = int(os.environ.get('ENCOPUS', True))
S3 = int(os.environ.get('S3', False))

logging.basicConfig(level=LOG_LEVEL)
log = logging.getLogger('KZDROE')

app = Flask(__name__)
metrics = UWsgiPrometheusMetrics(app, group_by='endpoint')
metrics.info('app_info', 'Application info', version=__version__)
metrics.register_endpoint('/metrics')
metrics.start_http_server(METRICS_PORT)

if S3:
    from kzdroe.s3 import upload_file

@app.route('/ping', methods=['GET'])
def ping():
    return 'Pong', 200

@app.route('/kzr/<accountid>/<userid>/<recfile>', methods=['PUT'])
@metrics.counter(
    'cnt_accountid', 'Number of invocations per accountid', labels={
        'accountid': lambda: request.view_args['accountid'],
        'status': lambda resp: resp.status_code
    })
@metrics.counter(
    'cnt_userid', 'Number of invocations per userid', labels={
        'userid': lambda: request.view_args['userid'],
        'status': lambda resp: resp.status_code
    })
def upload(accountid, userid, recfile):
    result={}
    now_greg_timestamp = datetime.now().timestamp() + GREGORIAN_SECONDS
    start_greg_timestamp = request.args.get(
        'start', 
        default = now_greg_timestamp, 
        type = int
    )
    result['start_greg_timestamp']=start_greg_timestamp
    calldatetime=datetime.fromtimestamp(
        start_greg_timestamp - GREGORIAN_SECONDS
    )
    result['calldatetime']=calldatetime
    dstpath = (
        "rec/" + 
        accountid + 
        "/" + 
        userid + 
        calldatetime.strftime("/%Y/%m/%d/%H")
    )
    dstfile = dstpath + "/" + recfile
    result['srcfile']=recfile
    result['dstfile']=dstfile
    if not os.path.exists(dstpath):
        os.makedirs(dstpath)
    with open(dstfile, 'wb') as f:
        print(dir(request))
        f.write(request.data)
    if ENCOPUS:
        opusdstfile = os.path.splitext(dstfile)[0] + ".opus"
        log.info(
            'running: opusenc --speech {} {}'.format(
                dstfile, 
                opusdstfile
            )
        )
        if Popen(
            [
                'opusenc',
                '--speech', 
                dstfile, 
                opusdstfile 
            ]).wait() == 0:
            log.info(
                'converted {} -> {}'.format(
                    dstfile, 
                    opusdstfile
                )
            )
            finalfile=opusdstfile
            result["opusfile"]=opusdstfile
            result["finalfile"]=opusdstfile
            result['opus_transcode']=True
    else:
        finalfile=dstfile
        result["finalfile"]=dstfile
        result['opus_transcode']=False
    if S3:
        if upload_file(finalfile,accountid):
            result["s3_key"]="s3://" + accountid + "/" + finalfile
            result["s3_upload"]=True
        else:
            result["s3_upload"]=False
    return jsonify(result)

if __name__ == '__main__':
    if not (S3_ACCESS_KEY and S3_SECRET_KEY):
        log.error(
            "ERROR: for S3 storage need S3_ACCESS_KEY and S3_SECRET_KEY"
        )
        exit(1)
    if ENCOPUS:
        if which('opusenc') is None:
            log.error("ERROR: opusenc not found in PATH - please make sure it's installed")
            exit(1)
    metrics.start_http_server(METRICS_PORT)
    app.run(host='0.0.0.0')